price-estimator/src/solver_rounding_buffer.rs

```diff
@@ -0,0 +1,162 @@
+use core::models::{AccountState, Order, TokenId, TokenInfo};
+use ethcontract::{Address, U256};
+use std::collections::{BTreeMap, HashMap};
+
+// This code is closely related to dex-solver/src/opt/process/Rounding.py .
+// Discussion of motivation happened in https://github.com/gnosis/dex-services/issues/970 .
+
+type Tokens = BTreeMap<TokenId, TokenInfo>;
+
+const MAX_ROUNDING_VOLUME: f64 = 100_000_000_000.0;
+const PRICE_ESTIMATION_ERROR: f64 = 10.0;
+
+fn max_rounding_amount(token_price: f64, fee_token_price: f64) -> f64 {
+    let estimated_price_in_fee_token = token_price / fee_token_price;
+    let max_rounding_amount = MAX_ROUNDING_VOLUME / estimated_price_in_fee_token;
+    max_rounding_amount.max(1.0)
+}
+
+/// Perform the same rounding buffer calculation as our solvers in order to increase the correctness
+/// of our estimates.
+/// All token prices must be nonzero.
+#[allow(dead_code)]
+pub fn apply_rounding_buffer(
+    tokens: &Tokens,
+    orders: &mut Vec<Order>,
+    account_state: &mut AccountState,
+) {
+    let fee_token_price = tokens
+        .get(&TokenId(0))
+        .expect("fee token does not have price")
+        .external_price as f64;
+    assert!(fee_token_price > 0.0, "fee token price is 0");
+
+    // The maximum rounding buffer over all orders from this address selling this token.
+    let mut account_balance_buffers = HashMap::<(Address, TokenId), u128>::new();
+    // Apply rounding buffer to account balances and order sell amounts.
+    for o in orders.iter_mut() {
+        // Compute rounding buffer.
+        let (t_s, t_b) = (TokenId(o.sell_token), TokenId(o.buy_token));
+        let estimated_buy_token_price = match tokens.get(&t_b) {
+            Some(TokenInfo { external_price, .. }) if *external_price > 0 => *external_price as f64,
+            _ => continue,
```

Comment:
Does the solver just not add a rounding buffer for tokens without estimates? I imagine it still needs some rounding buffer for these tokens.

Reply:
Ah I meant to have a comment here about how we should handle missing prices. The solver calculates the prices in a more complicated way through an optimization model that the estimated prices are part of so it's not clear to me what we should do because we can't implement that optimization model.
I'll ask in the issue about it.

Reply:
We should have price estimates for all tokens that are already connected to the fee token. 
I'm not entirely sure what happens if we have an order that is not yet connected to the fee token, but the price estimator is using it to suggest an order that will make it connected. But then this is such an exotic case that we can probably ignore it and not apply any rounding in that case.